### DISCUSSION TITLE
The `IndexCreateStatement` does not work with mysql, if the name is not provided

### DISCUSSION THREAD

**Discussion by mohs8421** [SeaQL/sea-orm#1620] [Ideas]
## Description
When an `IndexCreateStatement` is not build with a name, the result in mysql will look like this:
```sql
CREATE INDEX  ON `table` (`col`)
```
That query will fail, because of the missing name.

## Steps to Reproduce
1. Write an IndexCreateStatement, ommitting the name
2. Try to execute it

## Suggested Solution
Use builder pattern to expect the `name`, `table` and `col` properties of the `IndexCreateStatement` to be present. See: https://crates.io/crates/derive_builder

### Expected Behavior
It would be better if the creation of the `IndexCreateStatement` fails directly, best at compile time, so you cannot forget it. Same goes to providing a table and at least one column.

### Actual Behavior
It fails at runtime when the query should be executed

### Reproduces How Often
always

### Workarounds
Remember to put everything in the first place

**Comment by tyt2y3** [comment]
Thank you for the suggestion. Does `derive_builder` support mandatory field?
I think such an API is possible, with a more complex type system. Basically, `IndexBuilder` -> `IndexWithName` -> `IndexWithNameAndTable` etc? I am worried about the compilation time / complexity with this approach.

**Reply by mohs8421** [reply]
I think it would be simpler to have the `build()` function return a result, as in the example of the `derive_builder` crate.
Yes, that would cause it to not fail at compile time, but at runtime, but at least before execution in sql.
I'm not sure whether this can be pushed to fail at compile time without writing a macro for that.

**Comment by billy1624** [comment]
I do remember there are builder pattern which enforce the checking at compile time. Here is one of them https://crates.io/crates/typed-builder

**Reply by mohs8421** [reply]
this looks like the best matching solution, might even be valuable for other use cases like other common statements.

**Reply by billy1624** [reply]
Hey @mohs8421, you wanna give it a try? :P

**Reply by mohs8421** [reply]
I would, but have quite some other issues that consume my time

**Reply by mohs8421** [reply]
I am currently playing with that, although I think regarding this current issue, I go back to a different approach - which is generating a name from the given information.

**Comment by billy1624** [comment]
Context for newcomers, related PR:
- https://github.com/SeaQL/sea-query/pull/638
